File: packages/gbnf-experiment/data/20260909T123239Z_5982b634/ported_implementation/src/grammar-graph/pointers.ts
import type { GraphPointer, GraphPointerKey } from './graph-pointer.js';

export class Pointers {
  #pointers = new Map<GraphPointerKey, GraphPointer>();

  constructor(...pointers: GraphPointer[]) {
    for (const pointer of pointers) {
      this.add(pointer);
    }
  }

  add(pointer: GraphPointer): void {
    this.#pointers.set(pointer.id, pointer);
  }

  [Symbol.iterator](): IterableIterator<GraphPointer> {
    return this.#pointers.values();
  }

  get size(): number {
    return this.#pointers.size;
  }
}

File: proposed-projects/round-trip-experiment/reverse/20260909T175143Z_62b61540/20260910T094302Z_8f8be025/ported_implementation/gbnf/grammar_graph/pointers.py
from typing import Iterator

from .graph_pointer import GraphPointer


class Pointers:
    def __init__(self, *pointers: GraphPointer):
        self._pointers: dict[str, GraphPointer] = {}
        for pointer in pointers:
            self.add(pointer)

    def add(self, pointer: GraphPointer) -> None:
        self._pointers[pointer.id] = pointer

    def __iter__(self) -> Iterator[GraphPointer]:
        yield from list(self._pointers.values())

    @property
    def size(self) -> int:
        return len(self._pointers)

    def __len__(self) -> int:
        return len(self._pointers)
